### ISSUE TITLE
[1.12 rc4]service discovery failed

### ISSUE THREAD

**Issue by upccup** [moby/moby#24679]
now i create two service and network like this

```
   docker network create --driver overlay mynet
   docker service create --network mynet --name redis redis
   docker service create --network mynet --name web my-web
```

now i try to connect redis server in my web server but failed

```
➜  ~ docker exec -it  web.1.4i3r3f3djfgpwmtkc6kivacvf  sh
# ping redis
PING redis (10.0.0.2): 56 data bytes
92 bytes from f8c0f07ea3e3 (10.0.0.5): Destination Host Unreachable
92 bytes from f8c0f07ea3e3 (10.0.0.5): Destination Host Unreachable
92 bytes from f8c0f07ea3e3 (10.0.0.5): Destination Host Unreachable
92 bytes from f8c0f07ea3e3 (10.0.0.5): Destination Host Unreachable
```

because the redis container ip is changed

```
➜  ~ docker exec -it redis.1.6y9v4qwgd68b1e4b04khxl754 sh
# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
68: eth0@if69: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default
    link/ether 02:42:0a:00:00:03 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.3/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:aff:fe00:3/64 scope link
       valid_lft forever preferred_lft forever
70: eth1@if71: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:ac:12:00:03 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.3/16 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:3/64 scope link
       valid_lft forever preferred_lft forever
```

**docker version**

```
➜  ~ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: docker-253:0-34333044-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 959.6 MB
 Data Space Total: 107.4 GB
 Data Space Available: 17.27 GB
 Metadata Space Used: 2.425 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Swarm: active
 NodeID: 3q9rse4k6v0jlf12hb68fuelr
 IsManager: No
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.703 GiB
Name: localhost
ID: PD5B:R7W4:Q64C:X5VW:ET6Q:OFQP:GFYE:5NYG:V3DU:XS5Q:B6UL:RWIV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 42
 Goroutines: 101
 System Time: 2016-07-15T05:15:55.033930206-04:00
 EventsListeners: 1
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**service redis**

```
➜  ~ docker service inspect redis
[
    {
        "ID": "8jiozk9zwwm3ujry15kgn0alh",
        "Version": {
            "Index": 224
        },
        "CreatedAt": "2016-07-15T08:33:11.804914778Z",
        "UpdatedAt": "2016-07-15T08:33:11.808655794Z",
        "Spec": {
            "Name": "redis",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "redis"
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {}
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {},
            "Networks": [
                {
                    "Target": "28om24odrwbevea8kuwww7kpb"
                }
            ],
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip"
            },
            "VirtualIPs": [
                {
                    "NetworkID": "28om24odrwbevea8kuwww7kpb",
                    "Addr": "10.0.0.2/24"
                }
            ]
        }
    }
]
```

**service web**

```
➜  ~ docker service inspect web
[
    {
        "ID": "9zx2r9jpxrb04vsx8mdn10oe2",
        "Version": {
            "Index": 233
        },
        "CreatedAt": "2016-07-15T08:35:29.858726294Z",
        "UpdatedAt": "2016-07-15T08:35:29.86259807Z",
        "Spec": {
            "Name": "web",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "demoregistry.dataman-inc.com/library/yaoyun-web:0711"
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {}
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {},
            "Networks": [
                {
                    "Target": "28om24odrwbevea8kuwww7kpb"
                }
            ],
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip"
            },
            "VirtualIPs": [
                {
                    "NetworkID": "28om24odrwbevea8kuwww7kpb",
                    "Addr": "10.0.0.4/24"
                }
            ]
        }
    }
]
```

**Comment by hvpareja** [comment]
Hi @upccup;

First of all, I'd say the title of your issue is wrong. Service discovery works in your case as you're getting the container ip address via the docker embed DNS server. The problem should be elsewhere, but there is definitely a problem.

I have to say I have been experiencing the same error. But also with the same service. Service discovery works for me for other services (php-fpm, nginx, mailcatcher) but not for redis.

I wonder if there is some sort of default configuration in redis container that make the connection to fail.

Looking forward to read what the experts have to say in this regard.

Thank for your post!

**Comment by cpuguy83** [comment]
How did you init/join the cluster? Did you specify a `--listen-addr`?

**Comment by upccup** [comment]
@cpuguy83  yes i was init and join cluster with this conmand 
```
docker swarm init --listen-addr 192.168.59.103:2377
```
and
```
 docker swarm join  \
   --secret 9781uimms4epyyzzfhgc8shr1 \
   --ca-hash  sha256:a46be44f491e08afe0a1c59c136a25668df1c3c012338f3a0c7fe6d5a27872ec \
   --listen-addr 192.168.59.104:2377   192.168.59.103:2377
```

**Comment by upccup** [comment]
hi @hvpareja  misunderstand service discovery of new swarm with 1.12rc. but if this is not the right way to do service discovery how can id do this.

moreover ithink this problem was case by the service network not update when the container ip was changed

**Comment by hvpareja** [comment]
Hi @upccup. As you know, I was dealing with apparently the same issue. Then, I realised that one of my worker nodes had a different revision of the docker engine. As you surely know, you can check the version metadata with the command: `docker version`.

Once I updated the outdated node and recreated the whole swarm cluster, the overlay network worked internally with no issues. However, I now can't access to the published services from outside the docker network; I have the feeling it's something related with a miss-configured iptables. Though this is another issue I just wanted to share it.

I haven't said you're doing wrongly, by the way. I wouldn't dare as I don't consider I'm an expert to criticise others. I just pointed that your title doesn't describe your issue. You have a problem with the routing, not with the DNS resolution.

Let us know about the consistency of the docker version across your cluster.

Regards.

**Comment by upccup** [comment]
@hvpareja I'm so sorry of some may allow you to self-critical remarks that because of my bad english.

this is the docker version of the cluster

**swarm manager**
```
➜  ~ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: docker-253:0-34333044-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 954 MB
 Data Space Total: 107.4 GB
 Data Space Available: 17.21 GB
 Metadata Space Used: 2.318 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host overlay null
Swarm: active
 NodeID: dghvvy08ix9nbdy88nctadwwi
 IsManager: Yes
 Managers: 1
 Nodes: 2
 CACertHash: sha256:a46be44f491e08afe0a1c59c136a25668df1c3c012338f3a0c7fe6d5a27872ec
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.703 GiB
Name: localhost
ID: PD5B:R7W4:Q64C:X5VW:ET6Q:OFQP:GFYE:5NYG:V3DU:XS5Q:B6UL:RWIV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 52
 Goroutines: 151
 System Time: 2016-07-15T16:22:43.855599467-04:00
 EventsListeners: 1
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

**swarm worker**
```
➜  ~ docker info
Containers: 3
 Running: 2
 Paused: 0
 Stopped: 1
Images: 2
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: docker-253:0-34333044-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 970.5 MB
 Data Space Total: 107.4 GB
 Data Space Available: 17.26 GB
 Metadata Space Used: 2.527 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host overlay bridge
Swarm: active
 NodeID: 3q9rse4k6v0jlf12hb68fuelr
 IsManager: No
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.703 GiB
Name: localhost
ID: PD5B:R7W4:Q64C:X5VW:ET6Q:OFQP:GFYE:5NYG:V3DU:XS5Q:B6UL:RWIV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 52
 Goroutines: 119
 System Time: 2016-07-15T16:21:52.198587549-04:00
 EventsListeners: 2
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: true
Insecure Registries:
 127.0.0.0/8
```
**swarm node list**
```
➜  ~ docker node ls
ID                           HOSTNAME        MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
3q9rse4k6v0jlf12hb68fuelr    192.168.59.104  Accepted    Ready   Active
dghvvy08ix9nbdy88nctadwwi *  192.168.59.103  Accepted    Ready   Active        Leader
```

and about the iptables i have add all ports in 10-60000 with tcp/udp into firewall

**Comment by upccup** [comment]
@hvpareja  may this log of docker  deamon has some help
```
7月 15 16:42:10 localhost dockerd[3313]: time="2016-07-15T16:42:10.177723179-04:00" level=debug msg="starting exec command 03d00c71c7ba602c66f5b644dc16da1688decdd05983275cc46813b63067433e in container 1a33144b3136bc307052b7c8cad8d94a0b5e893c8376216b86f8960c46bc2827"
7月 15 16:42:10 localhost dockerd[3313]: time="2016-07-15T16:42:10.178831105-04:00" level=debug msg="Calling POST /v1.24/exec/03d00c71c7ba602c66f5b644dc16da1688decdd05983275cc46813b63067433e/resize?h=68&w=272"
7月 15 16:42:10 localhost dockerd[3313]: time="2016-07-15T16:42:10.178887900-04:00" level=debug msg="attach: stdin: begin"
7月 15 16:42:10 localhost dockerd[3313]: time="2016-07-15T16:42:10.178978227-04:00" level=debug msg="attach: stdout: begin"
7月 15 16:42:10 localhost dockerd[3313]: time="2016-07-15T16:42:10.250269931-04:00" level=debug msg="received containerd event: &types.Event{Type:\"start-process\", Id:\"1a33144b3136bc307052b7c8cad8d94a0b5e893c8376216b86f8960c46bc2827\", Status:0x0, Pid:\"03d00c71c7ba602c
7月 15 16:42:10 localhost dockerd[3313]: time="2016-07-15T16:42:10.250907482-04:00" level=debug msg="event unhandled: type:\"start-process\" id:\"1a33144b3136bc307052b7c8cad8d94a0b5e893c8376216b86f8960c46bc2827\" pid:\"03d00c71c7ba602c66f5b644dc16da1688decdd05983275cc4681
7月 15 16:42:11 localhost dockerd[3313]: time="2016-07-15T16:42:11.948577261-04:00" level=debug msg="2016/07/15 16:42:11 [DEBUG] memberlist: TCP connection from=192.168.59.104:37410\n"
7月 15 16:42:16 localhost dockerd[3313]: time="2016-07-15T16:42:16.593211671-04:00" level=debug msg="Name To resolve: redis."
7月 15 16:42:16 localhost dockerd[3313]: time="2016-07-15T16:42:16.593261132-04:00" level=debug msg="Lookup for redis.: IP [10.0.0.2]"
7月 15 16:42:16 localhost dockerd[3313]: time="2016-07-15T16:42:16.593495039-04:00" level=debug msg="miss notification for dest IP, 10.0.0.2"
7月 15 16:42:16 localhost dockerd[3313]: time="2016-07-15T16:42:16.593544540-04:00" level=error msg="could not resolve peer \"10.0.0.2\": could not resolve peer: serf instance not initialized"
7月 15 16:42:17 localhost dockerd[3313]: time="2016-07-15T16:42:17.593585766-04:00" level=debug msg="miss notification for dest IP, 10.0.0.2"
7月 15 16:42:17 localhost dockerd[3313]: time="2016-07-15T16:42:17.593644882-04:00" level=error msg="could not resolve peer \"10.0.0.2\": could not resolve peer: serf instance not initialized"
7月 15 16:42:18 localhost dockerd[3313]: time="2016-07-15T16:42:18.595962016-04:00" level=debug msg="miss notification for dest IP, 10.0.0.2"
7月 15 16:42:18 localhost dockerd[3313]: time="2016-07-15T16:42:18.596015556-04:00" level=error msg="could not resolve peer \"10.0.0.2\": could not resolve peer: serf instance not initialized"
```

**Comment by thaJeztah** [comment]
Looks like this may be relevant;

```
could not resolve peer: serf instance not initialized
```

**Comment by upccup** [comment]
@thaJeztah  but i think this error is cause thy wrong ip

**Comment by outcoldman** [comment]
I found a way how to easily get into state of "could not resolve peer: serf instance not initialized". Create 3 docker-machine instances, initialize swarm mode on one of them, create network=overlay also on this node, after that join another two nodes - you will get into state, when docker on other two nodes will report `serf instance not initialized`. If you will create 3 docker-machine instances, initialize and join the swarm - and only after that create network=overlay - everything seems working.

My `Makefile` (working).

```
swarm-setup:
	docker-machine create \
		--driver=virtualbox \
		--virtualbox-disk-size=200000 \
		--virtualbox-memory=512 \
		--virtualbox-cpu-count=1 \
		registry
	eval $$(docker-machine env registry) && docker run \
		--name registry \
		--publish 80:5000/tcp \
		-d registry:2
	docker-machine create \
		--driver=virtualbox \
		--virtualbox-disk-size=200000 \
		--virtualbox-memory=4096 \
		--virtualbox-cpu-count=2 \
		--engine-insecure-registry $$(docker-machine ip registry) \
		splunk1
	docker-machine create \
		--driver=virtualbox \
		--virtualbox-disk-size=200000 \
		--virtualbox-memory=3072 \
		--virtualbox-cpu-count=2 \
		--engine-insecure-registry $$(docker-machine ip registry) \
		splunk2
	docker-machine create \
		--driver=virtualbox \
		--virtualbox-disk-size=200000 \
		--virtualbox-memory=3072 \
		--virtualbox-cpu-count=2 \
		--engine-insecure-registry $$(docker-machine ip registry) \
		splunk3
	eval $$(docker-machine env splunk1) && docker swarm init --listen-addr $$(docker-machine ip splunk1) --secret splunk_swarm_cluster
	@echo "Do (with --ca-hash from above): "
	@echo "eval $$(docker-machine env splunk2) && docker swarm join --listen-addr $$(docker-machine ip splunk2) --secret splunk_swarm_cluster --ca-hash HASH_HERE $$(docker-machine ip splunk1):2377" 
	@echo "eval $$(docker-machine env splunk3) && docker swarm join --listen-addr $$(docker-machine ip splunk3) --secret splunk_swarm_cluster --ca-hash HASH_HERE $$(docker-machine ip splunk1):2377"
	@echo "eval $$(docker-machine env splunk1) && docker network create --driver=overlay splunk"
```

To make it not working you need to invoke ` docker network create --driver=overlay splunk` right after `docker swarm init`.

**Comment by outcoldman** [comment]
Actually still getting these errors

**Comment by hvpareja** [comment]
Could this issue:

https://github.com/docker/docker/issues/25087

Be related? Neither too very conclusive anyway ...

**Comment by thaJeztah** [comment]
ping @mavenugo

**Comment by mavenugo** [comment]
@upccup @Ismael I just want to clarify that SD is working as expected. If the connectivity is failing, we have to figure out whats going on with overlay networking and load-balancing. That would require more information or clear reproduction steps so that we can see whats going on.

**Comment by outcoldman** [comment]
In 1.12 all issues with load balancing are gone, but the network discovery by hostname does not work, see https://github.com/docker/docker/issues/25236. It does not seem like the same bug as described here.

**Comment by jhorwit2** [comment]
@mavenugo @thaJeztah afaik this is expected because ping isn't supported by ipvs implementation currently.

**Comment by jhorwit2** [comment]
Service discovery seems fine to me though. I can both ping and curl services. 

```
docker service create --name swarmtest --replicas 2 --network backend jhorwit2/swarm-test:0.0.1
```

Then in another container:

 ```
$ docker exec -it (docker ps -lq)  bash
curl swarmtest
eth0:	10.0.0.10
root@b4184eab6c47:/# curl swarmtest
eth0:	10.0.0.9
root@b4184eab6c47:/# curl swarmtest
eth0:	10.0.0.10
root@b4184eab6c47:/# curl swarmtest
eth0:	10.0.0.9
```
As you can see i'm RR between each vip in the network for swarmtest service.

**Comment by mavenugo** [comment]
Thanks @jhorwit2 for confirming. and yes we added partial support for ping service-name due to popular demand.

anyways, I think we still need to figure out the specifics of issues seen in @upccup @Ismael environments.

**Comment by Ismael** [comment]
I updated to 1.12 and still have the same issue, I can't ping. But wget from one to the other does work. It didn't occur to me to try wget directly. Ping is the basic tool for network connectivity for me.

You should complete the docs. docker service create doesn't tell you how to mount things. In rc5 the parameter was "dest", but now it's "destination" (I found it by trial and error).

---

How to reproduce my environment

# Networking

In Debian's /etc/network/interfaces
```
auto lxcbr0
iface lxcbr0 inet dhcp
        bridge_ports eth1
        bridge_fd 0
        bridge_maxwait 0
```


# Vagrant File

```
# -*- mode: ruby -*-
# vi: set ft=ruby :

# All Vagrant configuration is done below. The "2" in Vagrant.configure
# configures the configuration version (we support older styles for
# backwards compatibility). Please don't change it unless you know what
# you're doing.
Vagrant.configure(2) do |config|
  config.vm.box = "debian/jessie64"

  # Create a public network, which generally matched to bridged network.
  # Bridged networks make the machine appear as another physical device on
  # your network.
  config.vm.network "public_network", bridge: "lxcbr0", use_dhcp_assigned_default_route: true

  config.vm.provider "virtualbox" do |vb|
     vb.memory = "1024"
  end

  config.vm.provision "shell", inline: <<-SHELL    
    sudo apt-get update
    sudo apt-get install htop
    sudo sh /vagrant/getdocker.sh #script from get docker
    sudo usermod -aG docker vagrant
    
    sudo echo "10.11.12.8:/fileserver/archivos/Ismael/docker /media/fileserver nfs rw 0 0" >> /etc/fstab
    sudo mount -a

  SHELL
  
  config.vm.define "d1" do |d1|
    config.vm.provision "shell", inline: <<-SHELL
       echo "d1" > /etc/hostname
    SHELL
  end

  config.vm.define "d2" do |d2|
    config.vm.provision "shell", inline: <<-SHELL
       echo "d2" > /etc/hostname
    SHELL
  end
  
  config.vm.define "d3" do |d3|
    config.vm.provision "shell", inline: <<-SHELL
       echo "d3" > /etc/hostname
    SHELL
  end
  
end
```

# Swarm creation
In my case, my dhcp assigned d1: 10.11.0.125, d2: 10.11.0.138, d3: 10.11.0.174

## For d1
```
vagrant ssh d1
docker swarm init --listen-addr 10.11.0.125  --advertise-addr 10.11.0.125
```
Copy and paste the how to join masters

## For d2
`docker swarm join --token TOKEN --advertise-addr 10.11.0.138 --listen-addr 10.11.0.138 10.11.0.125:2377`

## For d3
`docker swarm join --token TOKEN --advertise-addr 10.11.0.174 --listen-addr 10.11.0.174 10.11.0.125:2377`

# Net
I don't know if this is really necessary

`docker network create --driver=overlay mynet`

# Service creation
I used grafana and Influxdb because that was what I wanted to test, I guess it'd do the same with any other container

**Grafana**

`docker service create --name=grafana --network mynet -p 3000:3000 --mount type=bind,source=/media/fileserver/grafana,destination=/var/lib/grafana grafana/grafana`


**InfluxDB**

`docker service create --replicas 1 --network mynet --name influxdb -p 8083:8083 -p 8086:8086 -p 2004:2004 --mount type=bind,source=/media/fileserver/,destination=/media/fileserver/ --mount type=bind,source=/media/fileserver/influxdb.conf,destination=/etc/influxdb/influxdb.conf:ro  influxdb:0.13 -config /etc/influxdb/influxdb.conf`



---

After I recreated the VMs from scratch using 1.12.0, I can't ping or wget from machine to machine. The DNS does not resolve to anything.

Furthermore, now I can't access the exposed ports in ANY host. Before I could access the container port from the host it was running on (and only that one). Now, I can't access it from outside.

Thanks for any help!

**Comment by Ismael** [comment]
I redid all my install and now I can use the container port from the host it is running on. Trying to access on another master does not work.

**Comment by Ismael** [comment]
> I redid all my install and now I can use the container port from the host it is running on. Trying to access on another master does not work.

Changed the base os to CentOS 7. Here this works. Should I create a separate issue for this?

**Comment by upccup** [comment]
with docker 1.12 in CentOS 7 this problem has been solved

**Comment by thaJeztah** [comment]
@upccup did you update CentOS, or docker? What were you running before?

**Comment by upccup** [comment]
@thaJeztah  Yes i update docker from 1.12rc4 to 1.12

**Comment by thaJeztah** [comment]
Closing this issue, as it is resolved in 1.12.0

**Comment by Ismael** [comment]
It's still broken on Debian. Follow my setup.

**Comment by thaJeztah** [comment]
@Ismael sorry, I interpreted your previous comment as "it's resolved"; could you also provide `docker info` and `docker version` of a machine in your setup?

**Comment by Ismael** [comment]
I don't know what happened that it can't establish the leader
now  (probably dchp expired the leases, this was a test setup after all).

Anyhow, I'm personally not interested in this issue anymore. Docker swarm
is too early too consider for production. I'm gonna try Kubernetes instead.

My replication steps should suffice to help you reproduce the bug, as it
only depends on having vagrant running.
Hope that helps you, sorry.


'
vagrant@d1:~$ docker info
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 5
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: pending
 NodeID: 9qrv3etcy4970myhgnksmp3cz
 Error: rpc error: code = 2 desc = raft: no elected cluster leader
 Is Manager: true
 ClusterID:
 Managers: 0
 Nodes: 0
 Orchestration:
  Task History Retention Limit: 0
 Raft:
  Snapshot interval: 0
  Heartbeat tick: 0
  Election tick: 0
 Dispatcher:
  Heartbeat period: Less than a second
 CA configuration:
  Expiry duration: Less than a second
 Node Address: 10.11.0.14
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1000 MiB
Name: d1
ID: 6RMK:MRSS:V577:ZGMC:ONII:OBR3:GLNQ:7HLJ:IEWB:JF23:24ZX:UPXC
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Insecure Registries:
 127.0.0.0/8

'




'
vagrant@d1:~$ docker version
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:40:59 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:40:59 2016
 OS/Arch:      linux/amd64

'

**Comment by therealpadams** [comment]
Does this relate to the issue reported here?:
https://forums.docker.com/t/interconnecting-services-in-1-12-swarm-mode-nginx-in-front-of-elastic/18583

Crate.io has the same problem as Elastic, our nodes are incapable of communication with each other since 1.12 rc4 (and including 1.12.0). Exactly as is described in the forum post I have linked above.

**Comment by sam-thibault** [comment]
This is an old issue.  I will close it as stale.
